tests/integration/tls/helpers_manual_tls.py

```diff
@@ -0,0 +1,209 @@
+
+
+class CSRsMissingError(Exception):
+    """Exception raised when the number of CSRs in the queue is less than the expected number."""
+
+    def __init__(self, message: str) -> None:
+        """Initialise the exception."""
+        super().__init__(message)
+
+
+class ProvidingCertificateFailedError(Exception):
+    """Exception raised when providing a certificate fails."""
+
+    def __init__(self, message: str) -> None:
+        """Initialise the exception."""
+        super().__init__(message)
+
+
+class CSR(NamedTuple):
+    """CSR represents the information about a certificate signing request."""
+
+    relation_id: str
+    application_name: str
+    unit_name: str
+    csr: bytes
+    is_ca: bool
+
+    @classmethod
+    def from_charm_csr(cls, charm_csr: dict) -> "CSR":
+        """Create a CSR object from a dictionary.
+
+        Arguments:
+        ---------
+            charm_csr : dict
+                The dictionary containing the information about
+                the certificate signing request gotten from the charm.
+
+        Returns:
+        -------
+            CSR: The CSR object.
+
+        """
+        return cls(
+            relation_id=charm_csr["relation_id"],
+            application_name=charm_csr["application_name"],
+            unit_name=charm_csr["unit_name"],
+            csr=charm_csr["csr"].encode(),
+            is_ca=charm_csr["is_ca"],
+        )
+
+
+class ManualTLSAgent:
+    """An agent that processes certificate signing requests from the TLS operator."""
+
+    def __init__(self, tls_unit: Unit) -> None:
+        """Initialise the agent."""
+        self.tls_unit = tls_unit
+        self.ca_key = generate_private_key()
+        self.ca = generate_ca(self.ca_key, "CN_CA")
+        self.csr_queue = deque()
+
+    async def get_outstanding_certificate_requests(self) -> None:
+        """Get the outstanding certificate requests from the TLS operator.
+
+        Raises
+        ------
+            GettingOutstandingCertificateRequestsFailedError:
+                If getting the outstanding certificate requests fails.
+
+        """
+        logging.info("Getting outstanding certificate requests")
+        action = await self.tls_unit.run_action("get-outstanding-certificate-requests")
+        action: Action = await action.wait()
+        if action.status != "completed":
+            message = action.safe_data.get(
+                "message",
+                "Failed to get outstanding certificate requests",
+            )
+            if "No outstanding certificate requests" in message:
+                logging.info(message)
+            if "No certificates relation has been created yet" in message:
+                logging.info(message)
+            logging.error(message)
+            raise GettingOutstandingCertificateRequestsFailedError(message)
+        csrs = json.loads(action.results["result"])
+        self.csr_queue = deque([CSR.from_charm_csr(csr) for csr in csrs])
```

Comment:
question: why opt for a `deque` instead of a simple list? and use `pop(0)` instead of `popleft`

Reply:
A Python `list` is implemented as a dynamic array in C. If we use a list and pop(0) all subsequent items in the list must be shifted which renders the operation very inefficient. `deque` (short for double-ended queue) has O(1) performance for both append and pop on either side of the queue. [Resource](https://docs.python.org/3/library/collections.html#collections.deque)

Reply:
In general - in **tests** and with this limited scale, we optimize for readability more than we do for performances

Reply:
I understand. Since we do not care about the order of the CSRs (they can arrive in any order.) I will change it to a `list` and use `pop` to treat it as a stack rather than a queue. This way we get readability and still be efficient. WDYT?

Reply:
sounds good!